aten/src/ATen/test/tensor_interop_test.cpp

```diff
@@ -0,0 +1,140 @@
+#include "gtest/gtest.h"
+
+#include "ATen/ATen.h"
+#include <caffe2/core/init.h>
+#include <caffe2/core/operator.h>
+
+TEST(TestTensorInterop, Caffe2ToPytorchSimple) {
+  caffe2::Tensor c2_tensor(caffe2::CPU);
+  c2_tensor.Resize(4, 4);
+  auto data = c2_tensor.mutable_data<int64_t>();
+  for (int64_t i = 0; i < 16; i++) {
+    data[i] = i;
+  }
+  at::Tensor at_tensor(c2_tensor.getIntrusivePtr());
+  at::TensorImpl* impl = at_tensor.unsafeGetTensorImpl();
+
+  auto it = impl->data<int64_t>();
+  for (int64_t i = 0; i < 16; i++) {
+    ASSERT_EQ(it[i], i);
+  }
+}
+
+TEST(TestTensorInterop, Caffe2ToPytorchOp) {
+  caffe2::Tensor c2_tensor(caffe2::CPU);
+  c2_tensor.Resize(3, 3);
+  auto data = c2_tensor.mutable_data<int64_t>();
+  for (int64_t i = 0; i < 9; i++) {
+    data[i] = i;
+  }
+  at::Tensor at_tensor(c2_tensor.getIntrusivePtr());
+
+  ASSERT_EQ(at::sum(at_tensor).item<int64_t>(), 36);
+}
+
+TEST(TestTensorInterop, Caffe2ToPytorchUnsupportedDevice) {
+  caffe2::Tensor c2_tensor(caffe2::IDEEP);
+  at::Tensor at_tensor(c2_tensor.getIntrusivePtr());
+  ASSERT_ANY_THROW(at::sum(at_tensor));
+}
+
+TEST(TestTensorInterop, PytorchToCaffe2Op) {
+  caffe2::Workspace workspace;
+  caffe2::NetDef net;
+
+  {
+    auto* blob_a = BlobGetMutableTensor(workspace.CreateBlob("a"), caffe2::CPU);
+    auto at_tensor_a = at::ones({5, 5}, at::dtype(at::kFloat));
+    caffe2::Tensor c2_tensor_a(at_tensor_a.getIntrusivePtr());
+    blob_a->ResizeLike(c2_tensor_a);
+    blob_a->ShareData(c2_tensor_a);
```

Comment:
I suppose this is a valid thing to test for, but it's not what I expected given the test name (PytorchToCaffe2Op). The reason is that you're not actually passing a PyTorch tensor to Caffe2; instead, what you're doing is passing a Caffe2 tensor which happens to share data with a PyTorch tensor. If you wanted to actually use the PyTorch tensor directly, you'd stuff it into the blob/workspace directly, using e.g. BlobSetTensor. (I'm not saying it's wrong to test this, it's just unexpected!)